src/lib/storage/buffer/frame.cpp

```diff
@@ -0,0 +1,169 @@
+    return _state_and_version.compare_exchange_strong(old_state_and_version,
+                                                      update_state_with_same_version(old_state_and_version, 1));
+  }
+  return false;
+}
+
+bool Frame::try_lock_exclusive(Frame::StateVersionType old_state_and_version) {
+  DebugAssert(state(old_state_and_version) == UNLOCKED || state(old_state_and_version) == MARKED ||
+                  state(old_state_and_version) == EVICTED,
+              "Frame must be unlocked, marked or evicted to lock exclusive, instead: " +
+                  std::to_string(state(old_state_and_version)));
+  return _state_and_version.compare_exchange_strong(old_state_and_version,
+                                                    update_state_with_same_version(old_state_and_version, LOCKED));
+}
+
+bool Frame::unlock_shared() {
+  while (true) {
+    auto old_state_and_version = _state_and_version.load();
+    auto old_state = state(old_state_and_version);
+    Assert(old_state > 0 && old_state <= LOCKED_SHARED, "Frame must be locked shared to unlock shared.");
+    auto new_state = old_state - 1;
+    if (_state_and_version.compare_exchange_strong(old_state_and_version,
+                                                   update_state_with_same_version(old_state_and_version, new_state))) {
+      return new_state == Frame::UNLOCKED;  // return true if last shared lock was released
+    }
+  }
+}
+
+void Frame::unlock_exclusive() {
+  Assert(state(_state_and_version.load()) == LOCKED,
+         "Frame must be locked to unlock exclusive. " + std::to_string(state(_state_and_version.load())));
+  _state_and_version.store(update_state_with_increment_version(_state_and_version.load(), UNLOCKED),
+                           std::memory_order_release);
+}
+
+Frame::StateVersionType Frame::update_state_with_same_version(Frame::StateVersionType old_version_and_state,
+                                                              Frame::StateVersionType new_state) {
+  constexpr auto SHIFT = NUM_BITS - STATE_SHIFT;
+  static_assert(SHIFT == 16, "Shift must be 8.");
+  return ((old_version_and_state << SHIFT) >> SHIFT) | (new_state << STATE_SHIFT);
+}
+
+Frame::StateVersionType Frame::update_state_with_increment_version(Frame::StateVersionType old_version_and_state,
+                                                                   Frame::StateVersionType new_state) {
+  constexpr auto SHIFT = NUM_BITS - STATE_SHIFT;
+  static_assert(SHIFT == 16, "Shift must be 8.");
+  return (((old_version_and_state << SHIFT) >> SHIFT) + 1) | (new_state << STATE_SHIFT);
+}
+
+bool Frame::is_unlocked() const {
+  return state(_state_and_version.load()) == UNLOCKED;
+}
+
+std::ostream& operator<<(std::ostream& os, const Frame& frame) {
+  auto state = Frame::state(frame.state_and_version());
+  auto version = Frame::version(frame.state_and_version());
+  auto node_id = frame.node_id();
+  auto dirty = frame.is_dirty();
+
+  std::stringstream ss;
```

Comment:
Besides the naming and auto to stick here, why not directly writing to the passed stream?

Reply:
Multiple threads may write to the same stream. This would interleave the output.

Reply:
`osyncstream` might better though

Reply:
Isn't printing the info just meant for test and debugging purposes?
If we write many things concurrently/in a hot loop to stdout (which will be the main usage of printing anything), then we have other issues than mingled output. And if users plan to do so, it should be their call to ensure a nice output by using a secomnd stream I guess.

Reply:
Better idea! Thanks!